### PR TITLE
Rewrite yaml merger

### LINKED ISSUE #7556: Add inheritance support to weapons [$15]
Currently only actors can inherit, while weapons have to be fully duplicated even if you need only a single change.

This would probably allow to shrink the number of yaml lines for weapons quite a bit.

<bountysource-plugin>
---
There is a **[$15 open bounty](https://www.bountysource.com/issues/8986300-add-inheritance-support-to-weapons?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #8086: Negating double inherited traits does not work
If a trait is on the default (`^Helicopter` e.g.) and on the actor (`TRAN`) then simply negating it (with `-`) won't work. Tested with `LeavesHusk` on the TD `TRAN` (`LeavesHusk` is on `TRAN` but not on the default. Adding it to default also breaks the custom removal on a map).
Also noticeable with `TRAN` in #8048 and `Selectable`. [Another example.](https://gist.github.com/abcdefg30/0e6a3f33e099f5e0efd2)

Sorry if I'm too confusing. ([Here](https://gist.github.com/abcdefg30/bb8959d4c9b9a7640a50) is my IRC spam regarding this.)

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/13072932-negating-double-inherited-traits-does-not-work?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #8185: Inheriting DefaultVariant does not work in voices.yaml
```yaml
    ^DefaultVoice:
	Voices:
		Burned: dm01
		Zapped: igenzapa
	Variants: .WAV
	DefaultVariant: .wav

    SovietMCVVoice:
	Inherits: ^DefaultVoice
	Voices:
		Select: vmcssea, vmcsseb, vmcssec, vmcssed, vmcssee
		Move: vmcsmoa, vmcsmob, vmcsmoc, vmcsmod, vmcsmoe
		Attack: vmcsmoa, vmcsmob, vmcsmoc, vmcsmod, vmcsmoe`
```
used in voices.yaml results with
```
Attempt to Play2D a null `ISoundSource`
LoadSound, file does not exist: vmcssed.aud
Attempt to Play2D a null `ISoundSource`
LoadSound, file does not exist: vmcssec.aud
```
messages.

I don't think that's supposed to happen. The wavs are loose and bits reference lists their folder.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/15554630-inheriting-defaultvariant-does-not-work-in-voices-yaml?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #8890: Sequence inheritance MiniYaml merging [$15]
Like we do for rules as proposed in https://github.com/OpenRA/OpenRA/pull/8799#issuecomment-126936258.

<bountysource-plugin>
---
There is a **[$15 open bounty](https://www.bountysource.com/issues/25281380-sequence-inheritance-miniyaml-merging?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #10207: YAML merging issues
Since #9710, yaml rules that used to be fine in our missions and mini-games don't work anymore (see [irc logs](http://logs.openra.net/?year=2015&month=12&day=12#16:10:29), for example).

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29073937-yaml-merging-issues?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by pchote**
Fixes #10207
Fixes #7556
Fixes #8890
Fixes #8086
Fixes #8185

As I explained in #10207 all these bugs exist because the code tried to treat yaml merging and inheritance as separable steps (which is completely bogus). This moves the inheritance logic into the yaml parser and fixes the merge and removal logic to behave in the logical way.

### CODE REVIEW

File: OpenRA.Test/OpenRA.Game/MiniYamlTest.cs [hunk 1]
```diff
@@ -61,30 +44,6 @@ public class MiniYamlTest
 		Attribute1: Test
 ";
 
-		[TestCase(TestName = "Merging: mixed addition and removal")]
-		public void MergeYamlA()
-		{
-			var a = MiniYaml.FromString(mixedMergeA, "mixedMergeA");
-			var b = MiniYaml.FromString(mixedMergeB, "mixedMergeB");
-
-			// Merge order should not matter
```
Comment (pchote) [inline OpenRA.Test/OpenRA.Game/MiniYamlTest.cs]: I removed this test because it was bogus (merge order *does* matter).

**Comment by pchote** [comment]
I don't understand the thf and hijacker lint errors on bomber john and drop zone w.  Neither of those maps touch anything related to infantry.

**Comment by pchote** [comment]
Found the issue, wrote a unit test for it, then fixed it.

**Comment by reaperrr** [comment]
#8086 seems to be fixed.

#8185 is not fixed.
For this change in RA's voice.yaml
```yaml
    ^DefaultDeathVoices:
	Voices:
		Die: dedman1,dedman2,dedman3,dedman4,dedman5,dedman7,dedman8
		Burned: dedman10
		Zapped: dedman6
	DisableVariants: Die, Burned, Zapped

    GenericVoice:
	Inherits: ^DefaultDeathVoices
	Variants:
		allies: .v01,.v03
		england: .v01,.v03
		france: .v01,.v03
		germany: .v01,.v03
		soviet: .r01,.r03
		russia: .r01,.r03
		ukraine: .r01,.r03
	Voices:
		Select: await1,ready,report1,yessir1
		Action: ackno,affirm1,noprob,overout,ritaway,roger,ugotit
```

I got
> Exception of type `System.InvalidOperationException`: Can't find Die in voice pool.

**Comment by pchote** [comment]
@reaperrr: did you try that before or after my latest fixup?  The issue with `-IdleSequences` that I have just fixed would have also caused what you just described.

**Comment by reaperrr** [comment]
Before, works now.

**Comment by reaperrr** [comment]
Tested weapon and sequence inheritance, encountered no issues.

The code is out of my league, but the way it works in practice gets my :+1:

**Comment by obrakmann** [comment]
Looks fine to me. The only inconsistency I could find was this:

rules/aircraft.yaml:
```
U2:
        -Selectable:
```

in map.yaml:
```
U2:
        -Selectable:
```
doesn't throw an error for the double removal.

```
U2.copy:
      Inherits@u2: U2
      -Selectable:
```
does throw an error.

I guess that's because of [this call to `Union`](https://github.com/OpenRA/OpenRA/pull/10648/files#diff-e5e78d9e1c86242278d0cbad542276f5R352), which would merge all duplicate lines for an actor entry. I'm fine with that. Leaving the comment here for a bit to see if anyone else complains, otherwise this gets my :+1: as well.

**Comment by pchote** [comment]
Correct.  Merging all keys with the same identifier together is fundamental to the way our rules are set up.  I don't think there's any reason to handle removals differently.

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/83e2eab7e40f415e6c80c49d47164c656040ec29)

**Pull request merged**